### DISCUSSION TITLE
Detailed guide on how to install bare Axolotl on PinePhone Manjaro Phosh

### DISCUSSION THREAD

**Discussion by ijustmadeanaccountforthis** [axolotl-chat/axolotl#615] [Show and tell]
This is meant as a little help, to get more people to test and probably debug this awesome project.
This is an unofficial guide and may not work on your phone/OS. But it could help to get you started.

It was tested on a fresh install of Manjaro ARM Beta17 Phosh on the PinePhone.
If it works (with some tweaks) on another Phone/OS feel free to comment below.

Basically what we do is we:
   - install dependencies
   - build the project
   - install the libraries
   - and run the software.

**### Disclaimer: This is a mere guide to install the package. The current stage of development can be found in the [Issues](https://github.com/nanu-c/axolotl/issues) page. Every problem that occurs after the program has started, has nothing to do with this guide.**

# 1) Install dependencies
_Dependencies can be found [here](https://github.com/nanu-c/axolotl/blob/main/docs/INSTALL.md#bare) or at the top of the [Makefile](https://github.com/nanu-c/axolotl/blob/main/Makefile#L6) if you need a closer look.
A package which contains a missing binary can be found with `pacman -F /usr/bin/cargo` for example._

Update OS first
```
sudo pacman -Syu
```

## Install Go, Rust and cargo

```
sudo pacman -S go
sudo pacman -S rust
```

## Install node/npm
_We install node via nvm (node version manager) script as it is easier to switch between versions
and because it is the [arch way](https://wiki.archlinux.org/title/node.js)._
```
sudo pacman -S nvm

// load the nvm start script automatically
echo 'source /usr/share/nvm/init-nvm.sh' >> ~/.bashrc
source ~/.bashrc

// install node version specified in the repo
cd axolotl-web
nvm install
nvm use
```

## Install base-devel (build essentials package group)
_They might already be installed_

`sudo pacman -S base-devel`

# 2) Build the package
## Download Axolotl source code
```
mkdir -p ~/Documents/builds/Axolotl
cd ~/Documents/builds/Axolotl
git clone https://github.com/nanu-c/axolotl.git
```

## Build source code
_(follow instructions from github [README](https://github.com/nanu-c/axolotl#manually))_
### Build dependencies
```
cd ~/Documents/builds/Axolotl/axolotl
make build-dependencies
```

### Update browsers list if [necessary](https://github.com/nanu-c/axolotl#manually)
```
cd ~/Documents/builds/Axolotl/axolotl/axolotl-web
npx browserslist@latest --update-db
```

### Some hacks
Now before we actually build axolotl, axolotl-web, crayfish and zkgroup, we need some little hacks.
First, as [go](https://en.wikipedia.org/wiki/Go_(programming_language)) is relatively new and still undergoing critical changes, we need to add two environment variables to tell it to use it's old form.
_(Anyone interested can read more about it [here](https://maelvls.dev/go111module-everywhere/).)_

```
echo 'export GOPATH=/home/manjaro/go' >> ~/.bashrc
source ~/.bashrc
```
_Don't worry if the directory does not exist yet. It will be automatically created later._

To add the second environment variable (we need to append an environment variable before the `go get` command in `build-zkgroup:`) we can either manually edit the [Makefile](https://github.com/nanu-c/axolotl/blob/main/Makefile#L138)
```
cd ~/Documents/builds/Axolotl/axolotl
vim Makefile
```
   From: `$(GO) get -d github.com/nanu-c/zkgroup`
   To: `GO111MODULE=off $(GO) get -d github.com/nanu-c/zkgroup`

Or use this command to change every (currently 3) occurrence of "go get -d"in the Makefile to use the environment variable.
```
cd ~/Documents/builds/Axolotl/axolotl
sed -i 's/$(GO) get -d/GO111MODULE=off $(GO) get -d/' Makefile
```

Next, the gnu compiler collection is called `aarch64-unknown-linux-gnu-gcc` instead of `aarch64-linux-gnu-gcc` on
the PinePhone. Therefore we create a symlink to the compiler that we have:

```
sudo ln -s /usr/bin/aarch64-unknown-linux-gnu-gcc /usr/bin/aarch64-linux-gnu-gcc
```

### Build the packages
_This might take a while since it contains some sub calls as can be seen [here](https://github.com/nanu-c/axolotl/blob/main/Makefile#L43). Network connection is needed at multiple points in the script, so consider disabling automatic suspend._
```
cd ~/Documents/builds/Axolotl/axolotl
make build
```

# 3) Install the libraries

We need to move the zkgroup lib file to the correct place, where `make install` can find it.
```
make copy-zkgroup
```
Now we install the libraries we built before. (This requires a password multiple times)
```
make install
```

# 4) Run Axolotl
_Starting the application may take a few seconds. You've come this far, be a bit patient one last time ;)_
```
/usr/lib/crayfish &
cd /home/manjaro/Documents/builds/Axolotl/axolotl
LD_LIBRARY_PATH=/home/manjaro/Documents/builds/Axolotl/axolotl
/usr/bin/go run .
```
Crayfish is needed to run in the background.
The last three commands (to actually run axolotl) can be written in a shell script and called from everywhere.

# 5) Some final words
Once again I want to state that this guide does not guarantee that the app you just built is working. The signalgroup may change its libraries [without notice](https://github.com/signalapp/zkgroup#readme), which breaks this application. But there are many skilled people who work on fixing that in their spare time. The current state can be deduced from the [Issues](https://github.com/nanu-c/axolotl/issues) page of this github page.

Feel free to comment if some things change, or if you have a slightly modified version for another phone/OS.

The advantage of a bare install (compared to flatpak or snap) is that you can build the newest version as soon as it is in the repo. And in addition to that you learn something about the project, its setup and probably about your own system.

**Comment by olof-nord** [comment]
The two Makefile targets are adjusted to not depend on other targets: consequently the document is updated.

**Comment by olof-nord** [comment]
The nvm usage is also adjusted to make use of the .nvmrc file in the axolotl-web folder.
